W3

chlin1024 opened the pull request:
1. fix error for w1 and w2
2. add product, order, and order items entities
3. new features to create, read, search, update, and delete products
4. new features to create, read, search, update, and delete orders
5. add authorization guard for guest and admin
6. add pagination for users query

Commit:
add pagination feature to user query

Commit:
add new resource product. new features 1.create new product.2.get product by search 3.update product info 4.delete product by id.

Commit:
1.add order module. 2.add order and order items entity. 3.add feature to create new order.

Commit:
add features 1.delete order and order items by id 2. get order and order items by id 3. update order by id.

Commit:
1.add jwt Guard and authorization Guard. 2.add guard to auth , products, and orders route

Commit:
1.add guard to user routes

Commit:
add pagination and query search for orders

Commit:
1.set global validation pipe 2.use .env file for jwt sercet and database.

Commit:
add features to update order items

Commit:
1.remove password from user api return data 2.add guest module

alvisLu commented:
commit 格式可以參考 [約定式提交](https://www.conventionalcommits.org/zh-hant/v1.0.0/)。
可以利用 `類型`區分 commit 的層級。
這個也可以用在開分支上，因為這是練習我們先用 W1... 區分。

alvisLu commented on src/guest/guest.controller.ts:
```diff
@@ -0,0 +1,53 @@
+    @Param('id', ParseIntPipe) id: number,
+    @Req() { user },
+  ): Promise<Order> {
+    const order = await this.ordersService.getOrderById(id);
+    const orderGuestId = order.guest.id;
+    if (orderGuestId !== user.id) {
+      throw new UnauthorizedException('User ID does not match');
+    }
+    return order;
```
可以在 order service 或是 guest service 中建立 function，controller 只負責呼叫函式。

alvisLu commented on src/orders/dto/order.dto.ts:
```diff
@@ -0,0 +1,16 @@
+import { FinancialStatus, FulfillmentStatus } from '../orderStatus.enum';
+
+export class OrderDto {
+  admin: number;
```
dto 要用 class validate 檢查喔~

chlin1024 replied:
有在global level 用validation pipe檢查

alvisLu commented on src/orders/order.entity.ts:
```diff
@@ -0,0 +1,55 @@
+  @Column({ nullable: true })
+  note: string;
+
+  @Column()
```
可以用 @CreateDateColumn()

alvisLu commented on src/orders/orders.controller.ts:
```diff
@@ -0,0 +1,103 @@
+    return this.ordersService.getOrderByUserId(id);
+  }
+
+  @Put('update/:id')
```
RESTful API 的設計可以參考 [RESTful Web API 設計](https://learn.microsoft.com/zh-tw/azure/architecture/best-1. 1. practices/api-design) 

2. PATCH 和 PUT 雖然都可以用於更新，但它們也是有些差異。[HTTP Verbs: 談 POST, PUT 和 PATCH 的應用](https://ihower.tw/blog/archives/6483)
  - PUT: 新增或替換資料
  - PATCH: 部分更新
  以 order 為例，用 PUT 會以為是將指定的 order 整組資料更換，用 PATCH 則是更新部分資料。

alvisLu commented on src/orders/orders.service.ts:
```diff
@@ -0,0 +1,194 @@
+    if (orderBy) {
+      const cleanOrderBy = orderBy.replace(/^'|'$/g, '');
+      orderColumn = cleanOrderBy.split(':')[0];
+      orderType = cleanOrderBy.split(':')[1].toUpperCase() as 'ASC' | 'DESC';
```
可以參考 [Adding ORDER BY expression](https://github.com/typeorm/typeorm/blob/master/docs/select-query-builder.md#adding-order-by-expression) 讓排序可以更多樣

alvisLu commented on src/users/users.service.ts:
```diff
@@ -57,13 +86,13 @@ export class UsersService {
       throw new NotFoundException(`User ${userName} Not Found`);
     }
-    return result;
+    return omit(user, ['password']);
```
可以在 userEntity.password 加上 `@Column({
    nullable: true,
    select: false,
  })`
可以參考 
- [column-options](https://orkhan.gitbook.io/typeorm/docs/entities#column-options)
- [How to set a nullable database field to NULL with typeorm?](https://stackoverflow.com/questions/64635617/how-to-set-a-nullable-database-field-to-null-with-typeorm)

Commit:
1.set cors 2.set up logger

Commit:
1.fix orderDto issue 2.refator orders service

Commit:
refactor return password

Commit:
add database migration

Commit:
add create table migrations

Commit:
refactor routes

Commit:
fix(migrations): resolve database migrations issue

- Refactored directory structure for better organization.

Commit:
test(e2e): add sign-up and login e2e tests.

Commit:
seed database user

Commit:
1.fix seed database issue 2.a
dd user query e2e testing.

Commit:
seed product, order, order item table

Commit:
1.add product e2e testing 2.update user e2e testing to include authorization

Commit:
add auth logout e2e test

Commit:
add mailer

Commit:
clear and seed database before every e2e testing

Commit:
integrate email queue to send confirmation after order creation

Commit:
1.add new test to e2e testing 2.new feature send email upon sign up

Commit:
1.Fix issue of duplicate table error caused by e2e test synchronize excution. 2. Updated queue processing logic to manage failed jobs and set attempts times.

Commit:
Add - Cache Login session to authentication

Commit:
move login session caching to interceptor

Commit:
feature: use Event Emitter to notify user creation and send order creation email job

Commit:
use swagger to create api document

Commit:
Refactor event handling and logging

1. Moved event emitter for order creation and user registration to an interceptor.
2. Refactored login-session caching mechanism.
3. Added logger to capture errors in the mailer and failed jobs.

Commit:
1.Added endpoints for checkout and refund operation. 2.Implemented HTTP module integration with LinePay API for payment processing. 3.Updated database schema and migration scripts to include new tables, checkout_transaction, refund_transaction, and order_transaction.

Commit:
Alter Product and OrderItem tables to include sale_price, update order creation and checkout logic

Commit:
Implemented task scheduling for applying discounts and restore price to all products at a specified time.

Commit:
revise line-pay service

Commit:
implemnt docker file and docker-compose

chlin1024 committed:
Update README.md

chlin1024 committed:
Update README.md

chlin1024 committed:
Update README.md

chlin1024 committed:
Update README.md

chlin1024 committed:
Update README.md

Signed-off-by: Phoebe C Lin <40210128+chlin1024@users.noreply.github.com>

Commit:
create .env.example file

Commit:
update README.md

Commit:
update README with system architecture image.

Commit:
Integrate line pay to checkout proccess

Commit:
update README.md and .env.example file

Commit:
update README.md

chlin1024 committed:
Merge pull request #8 from chlin1024/w9

w9 implement docker file and docker-compose

chlin1024 committed:
Merge pull request #9 from chlin1024/w8

W8

chlin1024 committed:
Merge pull request #7 from chlin1024/w7

W7

chlin1024 committed:
Merge pull request #6 from chlin1024/w6

W6

chlin1024 committed:
Merge pull request #5 from chlin1024/w5

W5

chlin1024 committed:
Merge pull request #4 from chlin1024/w4

W4
